### PR TITLE
Feature #6918: Add option to adjust font size separately from GUI size.

### LINKED ISSUE #6918: Add option to adjust font size separately from GUI size.
With respect to #6482, I want to bring this back up for discussion
> **Reason for closing: _Not a bug_**
> > please use the GUI zoom setting to increase sizes

On modern high-density screens, we need to be able to set the font size separately from the overall GUI size. If I increase the GUI zoom, then all the buttons and boxes are too large and block my view of playing the game. However, with the GUI zoom at 1×, the text on the minimap is a mere 4 pixels! That's too small to be able to read comfortably, especially when it's that light dim grey over that orangy-brown box.

### PR DESCRIPTION
**Pull request by gregcarlin**
Adds an option in the "Game Options" next to "Interface Size" called "Font Size". Available options are normal, double, and quad. Tested with both freetype and sprite fonts. The only known hiccup is that I couldn't get sprite fonts to work properly past a complete "quad zoom". Eg. if the interface and font sizes are both set to double, the effective font size will be quad. Increasing either option further will not increase the font size. This works fine for freetype fonts though.

Thanks for taking a look! Super excited to be helping out, I love playing the game!

Fixes #6918

### CODE REVIEW

File: src/table/misc_settings.ini [hunk 1]
```diff
@@ -308,5 +308,14 @@ min      = ZOOM_LVL_MIN
 max      = ZOOM_LVL_OUT_4X
 cat      = SC_BASIC
 
+[SDTG_VAR]
+name     = ""font_zoom""
+type     = SLE_UINT8
+var      = _font_zoom
+def      = ZOOM_LVL_OUT_4X
```
Comment (GabdaZM) [inline src/table/misc_settings.ini:315]: The default value should be ZOOM_LVL_MIN, it was surprising to see everything so big on starting the game.

Comment (gregcarlin) [reply src/table/misc_settings.ini:315]: Good catch, thanks! Just updated.

File: src/lang/english.txt [hunk 2]
```diff
@@ -991,6 +991,13 @@ STR_GAME_OPTIONS_GUI_ZOOM_DROPDOWN_NORMAL                       :Normal
 STR_GAME_OPTIONS_GUI_ZOOM_DROPDOWN_2X_ZOOM                      :Double size
 STR_GAME_OPTIONS_GUI_ZOOM_DROPDOWN_4X_ZOOM                      :Quad size
 
+STR_GAME_OPTIONS_FONT_ZOOM                                      :{BLACK}Font size
+STR_GAME_OPTIONS_FONT_ZOOM_DROPDOWN_TOOLTIP                     :{BLACK}Select the interface font size to use
+
+STR_GAME_OPTIONS_FONT_ZOOM_DROPDOWN_NORMAL                      :Normal
+STR_GAME_OPTIONS_FONT_ZOOM_DROPDOWN_2X_ZOOM                     :Double size
+STR_GAME_OPTIONS_FONT_ZOOM_DROPDOWN_4X_ZOOM                     :Quad size
```
Comment (LordAro) [inline src/lang/english.txt:1003]: Pretty sure "size" is redundant here (and above). I'd also want to use the full "Quadruple"

File: src/script/api/game/game_window.hpp.sq [hunk 3]
```diff
@@ -1023,6 +1023,7 @@ void SQGSWindow_Register(Squirrel *engine)
 	SQGSWindow.DefSQConst(engine, ScriptWindow::WID_GO_RESOLUTION_DROPDOWN,                "WID_GO_RESOLUTION_DROPDOWN");
 	SQGSWindow.DefSQConst(engine, ScriptWindow::WID_GO_FULLSCREEN_BUTTON,                  "WID_GO_FULLSCREEN_BUTTON");
 	SQGSWindow.DefSQConst(engine, ScriptWindow::WID_GO_GUI_ZOOM_DROPDOWN,                  "WID_GO_GUI_ZOOM_DROPDOWN");
+	SQGSWindow.DefSQConst(engine, ScriptWindow::WID_GO_FONT_ZOOM_DROPDOWN,                  "WID_GO_FONT_ZOOM_DROPDOWN");
```
Comment (LordAro) [inline src/script/api/game/game_window.hpp.sq]: extra space. You should be able to generate changes to this file by running `src/script/api/generate_widget.sh` (on Linux)

**Review by LordAro — COMMENTED** [review]
Looks broadly fine in principle...

File: src/lang/english.txt (hunk 2 above)
Comment (gregcarlin) [reply src/lang/english.txt:1003]: I copied this from the GUI zoom options. Should I change those as well?

Comment (LordAro) [reply src/lang/english.txt:1003]: Oh so you did. Don't worry then, leave it

Comment (gregcarlin) [reply src/lang/english.txt:1003]: Sounds good, thanks!

File: src/script/api/game/game_window.hpp.sq (hunk 3 above)
Comment (gregcarlin) [reply src/script/api/game/game_window.hpp.sq]: That script doesn't seem to do anything for me (I'm running macOS) so I just removed the space manually.

**Comment by planetmaker** [comment]
Is it really intended, that the font size selection is relative to the interface size?

With the current PR, if I select 'double size interface size' and 'double size font size', I end up with 4x font size compared to both settings set to 'normal'.

As sprite fonts are the defaults, it should work for those as with non-sprite fonts, too.

**Comment by gregcarlin** [comment]
Without this PR, changing the interface size will also change the font size if the sprite font is used, but it will not affect the font size if a FreeType font is used. Are you proposing to change this behavior so that the interface size never affects the font size?

**Comment by PeterN** [comment]
@gregcarlin I would say exactly that behaviour should happen, as you have added the ability to (almost) independently set the font size.

This may need existing text strings to be altered to indicate that the existing selection is for graphical/icon elements and no longer affects text?

**Comment by gregcarlin** [comment]
Sounds good, I'll get on it!

File: src/script/api/game/game_window.hpp.sq (hunk 3 above)
Comment (glx22) [reply src/script/api/game/game_window.hpp.sq]: the script should work now

**Comment by gregcarlin** [comment]
I updated the behavior! I didn't change the strings because I wasn't sure what would be more clear and I think the current strings are fairly easy to understand. However, if someone has a specific recommendation I'd be more than happy to make the change.

File: src/zoom_type.h [hunk 4]
```diff
@@ -53,6 +53,8 @@ DECLARE_POSTFIX_INCREMENT(ZoomLevel)
 typedef SimpleTinyEnumT<ZoomLevel, byte> ZoomLevelByte;
 
 extern ZoomLevelByte _gui_zoom;
+extern ZoomLevelByte _font_zoom;
 #define ZOOM_LVL_GUI (_gui_zoom)
+#define ZOOM_LVL_FONT (2 - _font_zoom)
```
Comment (LordAro) [inline src/zoom_type.h]: why `2 - ` ? Add a comment?

File: src/settings_gui.cpp [hunk 5]
```diff
@@ -540,6 +559,13 @@ struct GameOptionsWindow : Window {
 				UpdateAllVirtCoords();
 				break;
 
+			case WID_GO_FONT_ZOOM_DROPDOWN:
+				GfxClearSpriteCache();
+				_font_zoom = (ZoomLevel) index;
```
Comment (LordAro) [inline src/settings_gui.cpp]: extra space after the cast

File: src/fontcache.cpp [hunk 6]
```diff
@@ -293,6 +293,10 @@ void FreeTypeFontCache::SetFontSize(FontSize fs, FT_Face face, int pixels)
 			pixels = Clamp(min(head->Lowest_Rec_PPEM, 20) + diff, scaled_height, MAX_FONT_SIZE);
 		}
 	}
+
+	/* Apply user-specified font zoom. */
+	pixels = pixels * (1 << _font_zoom);
```
Comment (LordAro) [inline src/fontcache.cpp]: `*=` ?

**Review by LordAro — DISMISSED** [review]

File: src/fontcache.cpp (hunk 6 above)
Comment (PeterN) [reply src/fontcache.cpp]: Or should it be one of (Un)ScaleByZoom here?

File: src/zoom_type.h [hunk 7]
```diff
@@ -53,6 +53,12 @@ DECLARE_POSTFIX_INCREMENT(ZoomLevel)
 typedef SimpleTinyEnumT<ZoomLevel, byte> ZoomLevelByte;
 
 extern ZoomLevelByte _gui_zoom;
+extern ZoomLevelByte _font_zoom;
 #define ZOOM_LVL_GUI (_gui_zoom)
+/*
+ * Invert _font_zoom to be consistent with _gui_zoom and ZOOM_LVL_GUI
+ * (higher numbers = smaller font)
+ */
```
Comment (PeterN) [inline src/zoom_type.h]: How is this consistent with _gui_zoom? Why doesn't ZOOM_LVL_GUI need this 2 - x?

**Comment by LordAro** [comment]
I spotted this at high font size levels. I suspect it's not directly related to your change, but it is revealed by it

**Comment by PeterN** [comment]
@LordAro there is code that changes the width, it probably also needs to be changed to be based on (and happen on) font zoom instead of gui zoom. May well be other instances.

File: src/zoom_type.h (hunk 7 above)
Comment (gregcarlin) [reply src/zoom_type.h]: `ZOOM_LVL_GUI` has the convention of its value being inversely correlated with the GUI size (a higher value indicates a smaller GUI). This seems to work well for how it's used, which I haven't looked _too_ much into but seems to me like a fixed array of preset sizes.
I had `_font_zoom` follow the same convention for consistency when setting the value but due to the way this value is ultimately used directly as a multiplier it has to be inverted somewhere. 2 is the maximum zoom size for both fonts and the GUI.

**Comment by gregcarlin** [comment]
@LordAro and @PeterN the issue seems to only appear if the font zoom is adjusted mid-game. If it's adjusted before a game is loaded the town names appear fine. This leads me to believe the issue is with timing (i.e. the width needs to be recalculated when the font zoom changes). I'll investigate further.

**Comment by gregcarlin** [comment]
Got it! Just needed to add a call to `UpdateAllVirtCoords`.

File: src/zoom_type.h (hunk 7 above)
Comment (PeterN) [reply src/zoom_type.h]: I think you either want to do it the same way as gui zoom is done, or don't use the ZoomLevel enum at all.

ZOOM_LVL_OUT_4X is for making everything smaller, but in your code you use it to mean 4x larger.

Comment (gregcarlin) [reply src/zoom_type.h]: Makes sense. I inverted it so it matches GUI zoom. I put the changes in a new commit so you could see them more easily, but I can squash the two commits together if you'd prefer.

**Review by PeterN — APPROVED** [review]

**Comment by gregcarlin** [comment]
Yay! Thanks!

**Pull request merged**